### ISSUE TITLE
BUG: paired t-test calculations

### ISSUE THREAD

**Issue by tylerjereddy** [lanl/ldrd_virus_work#32]
Based on some discussions with Nick re: LDRD manuscript, I've switched from independent testing to related/paired testing for our T-tests. A sample of the code changes I'm using locally, which should eventually make it into `main`, are below:

```diff
--- a/viral_seq/analysis/spillover_predict.py
+++ b/viral_seq/analysis/spillover_predict.py
@@ -832,13 +832,17 @@ def compare_workflow_aucs(
         print(f"{predictions_paths[i]}:")
         wf_files = sorted(glob.glob(os.path.join(predictions_paths[i], "*csv")))
         for file in wf_files:
+            if "ensemble" in file.lower() or "stacking" in file.lower():
+                # we only calculate statistics between non-ensemble classifiers
+                # for now
+                continue
             aucs[i] += get_aucs(file, dataset_files[i], target_columns[i])
         auc_mean = np.mean(aucs[i])
         auc_std = np.std(aucs[i])
         print(f"Workflow mean auc = {auc_mean:.3f} std = {auc_std:.3f}")
-    t_stat, p_val = stats.ttest_ind(aucs[0], aucs[1])
+    result = stats.ttest_rel(aucs[0], aucs[1])
     print("")
-    print(f"Student t-test: t_stat {t_stat:.3f}, p-value {p_val:.3f}")
+    print(f"T-test for 80 paired ROC AUC samples: statistic {result.statistic:.3f}, p-value {result.pvalue:e}")
```

A sample script that goes along with that to get the values printed out after the workflows are complete:

```python
from viral_seq.analysis import spillover_predict as sp

sp.compare_workflow_aucs(
            ("human_shuffled/data_calculated/predictions", "human_fixed/data_calculated/predictions"), 
            ("Relabeled_Test_Human_Shuffled.csv", "Relabeled_Test.csv"),
            #("Relabeled_Test_Human_Shuffled.csv", "Relabeled_Test_Mammal_Shuffled.csv"),
            #("Relabeled_Test_Primate_Shuffled.csv", "Relabeled_Test_Mammal_Shuffled.csv"),
            #("Mollentze_Holdout.csv", "Relabeled_Test_Mammal_Shuffled.csv"),
            #("Human Host", "mammal"),
            ("human", "human"),
            )


```

**Comment by tylerjereddy** [comment]
Nick is now suggesting that we use https://www.statsmodels.org/stable/generated/statsmodels.stats.descriptivestats.sign_test.html, since it is not parametric, doesn't assume normal distribution, and may provide more appropriate paired p-values for the ROC AUC comparisons.

The diff below is now being used to calculate the new p-values for ROC AUC comparisons:

```diff
--- a/viral_seq/analysis/spillover_predict.py
+++ b/viral_seq/analysis/spillover_predict.py
@@ -11,6 +11,7 @@ import matplotlib.pyplot as plt
 import pandas as pd
 import polars as pl
 import sklearn
+from statsmodels.stats.descriptivestats import sign_test
 from sklearn.ensemble import RandomForestClassifier
 from sklearn.model_selection import (
     StratifiedKFold,
@@ -832,13 +833,17 @@ def compare_workflow_aucs(
         print(f"{predictions_paths[i]}:")
         wf_files = sorted(glob.glob(os.path.join(predictions_paths[i], "*csv")))
         for file in wf_files:
+            if "ensemble" in file.lower() or "stacking" in file.lower():
+                # we only calculate statistics between non-ensemble classifiers
+                # for now
+                continue
             aucs[i] += get_aucs(file, dataset_files[i], target_columns[i])
         auc_mean = np.mean(aucs[i])
         auc_std = np.std(aucs[i])
         print(f"Workflow mean auc = {auc_mean:.3f} std = {auc_std:.3f}")
-    t_stat, p_val = stats.ttest_ind(aucs[0], aucs[1])
-    print("")
-    print(f"Student t-test: t_stat {t_stat:.3f}, p-value {p_val:.3f}")
+    print("Number of ROC AUCs in each condition:", len(aucs[0]), len(aucs[1]))
+    result = sign_test(np.asarray(aucs[0]) - np.asarray(aucs[1]), mu0=0)[1]
+    print(f"Sign test p-value for 80 paired ROC AUC samples: {result:.2e}")
```

**Comment by tylerjereddy** [comment]
Yet another change/improvement to p-value calculation based on discussions with Nick--we're now trying to only calculate p-values between different conditions over 10 random seeds at a time (once for each estimator). The reason for this is that we cannot expect independence between different estimators at the same random seed. So, we effectively do 10 ROC AUC from one estimator in one condition vs. the same estimator 10 ROC AUCs of another condition (for 10 seeds). Then we look at the min and max p values calculated that way across the 8 estimator types. Sample code diff for this calculation is below:

<details>

```diff
--- a/viral_seq/analysis/spillover_predict.py
+++ b/viral_seq/analysis/spillover_predict.py
@@ -11,6 +11,7 @@ import matplotlib.pyplot as plt
 import pandas as pd
 import polars as pl
 import sklearn
+from statsmodels.stats.descriptivestats import sign_test
 from sklearn.ensemble import RandomForestClassifier
 from sklearn.model_selection import (
     StratifiedKFold,
@@ -832,13 +833,28 @@ def compare_workflow_aucs(
         print(f"{predictions_paths[i]}:")
         wf_files = sorted(glob.glob(os.path.join(predictions_paths[i], "*csv")))
         for file in wf_files:
+            if "ensemble" in file.lower() or "stacking" in file.lower():
+                # we only calculate statistics between non-ensemble classifiers
+                # for now
+                continue
             aucs[i] += get_aucs(file, dataset_files[i], target_columns[i])
         auc_mean = np.mean(aucs[i])
         auc_std = np.std(aucs[i])
         print(f"Workflow mean auc = {auc_mean:.3f} std = {auc_std:.3f}")
-    t_stat, p_val = stats.ttest_ind(aucs[0], aucs[1])
-    print("")
-    print(f"Student t-test: t_stat {t_stat:.3f}, p-value {p_val:.3f}")
+    # we'd prefer to compare just the 10 seeds of the same estimator between
+    # two different conditions, since not all estimators are guaranteed to
+    # be independent at the same seed
+    n_estimators = 8
+    per_estimator_aucs_cond_1 = np.split(np.asarray(aucs[0]), n_estimators)
+    per_estimator_aucs_cond_2 = np.split(np.asarray(aucs[1]), n_estimators)
+    p_values = []
+    for estimator_ind in range(n_estimators):
+        p_values.append(sign_test(per_estimator_aucs_cond_1[estimator_ind]
+                        - per_estimator_aucs_cond_2[estimator_ind], mu0=0)[1])
+    min_p_val = np.min(p_values)
+    max_p_val = np.max(p_values)
+    print("all p vals:", p_values, "num p vals:", len(p_values))
+    print(f"Min and Max Sign test p-value for 10 paired ROC AUC samples by estimator type: {min_p_val:.3e}, {max_p_val:.3e}")
```

</details>
